sopel/modules/choose.py

```diff
@@ -12,8 +12,51 @@
 from __future__ import absolute_import, division, print_function, unicode_literals
 
 import random
+import unicodedata
 
-from sopel import plugin
+from sopel import formatting, plugin
+
+
+def _format_safe(text):
+    """Remove excess whitespace and terminate IRC formatting.
+
+    :param str text: text to clean of whitespace
+    :rtype: str
+
+    Our own take on ``str.strip()`` that skips stripping off IRC formatting
+    and makes sure any formatting codes are closed if necessary.
+    """
+    start = end = 0
+
+    pos = 0
+    while pos < len(text):
+        is_whitespace = unicodedata.category(text[pos]) == 'Zs'
+        is_non_printing = (
+            text[pos] in formatting.CONTROL_NON_PRINTING and
+            text[pos] not in formatting.CONTROL_FORMATTING
+        )
+        if not is_whitespace and not is_non_printing:
+            start = pos
+            break
+        pos += 1
+
+    pos = len(text) - 1
+    while pos >= 0:
+        is_whitespace = unicodedata.category(text[pos]) == 'Zs'
+        is_non_printing = (
+            text[pos] in formatting.CONTROL_NON_PRINTING and
+            text[pos] not in formatting.CONTROL_FORMATTING
+        )
+        if not is_whitespace and not is_non_printing:
+            end = pos + 1
+            break
+        pos -= 1
```

Comment:
For some reason I don't like `while` loop to iterate over a string, mostly because 1. you need to be sure of your exit condition and 2. you can already iterate over a string like any other iterable.

So I started playing with `enumerate`, to get both the current character and its position, played a bit with `reversed` and text's length to do that from the end of the string, and _voilà_, no more `while` loop, no more `pos` to remember, which means no risk of an off by one error, or a missing exit condition.

```suggestion
    # strip left
    for i, c in enumerate(text):
        is_whitespace = unicodedata.category(c) == 'Zs'
        is_non_printing = (
            c in formatting.CONTROL_NON_PRINTING and
            c not in formatting.CONTROL_FORMATTING
        )
        if not is_whitespace and not is_non_printing:
            start = i
            break

    # strip right
    for i, c in enumerate(reversed(text)):
        is_whitespace = unicodedata.category(c) == 'Zs'
        is_non_printing = (
            c in formatting.CONTROL_NON_PRINTING and
            c not in formatting.CONTROL_FORMATTING
        )
        if not is_whitespace and not is_non_printing:
            end = len(text) - i
            break
```

I checked with `timeit` with different string, and there is not a lot of difference: sometimes it's slightly faster, sometimes it's slightly slower.

Reply:
I'm going to steal the comments from this, at least.

For longer strings the `enumerate()` approach was consistently ~10-15% slower than `while` loops in my own `timeit` experiments from before this suggestion existed—significant enough to push me toward manually keeping track of `pos`. And even ignoring execution time, wrapping a generator inside another generator in the right-stripping code using `enumerate(reversed())` feels like unnecessary overhead.

Admittedly, there _was_ an off-by-one error in this code while I was first developing it, but that's why I wrote tests. Needing to be sure of the exit condition isn't a bad thing.

Reply:
I'm not surprised by the off-by-one error. It's a classic.

Also I don't think this function will be modified anytime soon, I just hope no one will ever want to do it because `while` loops are evil.